src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs

```diff
@@ -0,0 +1,238 @@
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.FederatedPeg.Features.FederationGateway
+{
+    /// <summary></summary>
+    public interface ITemplateTransactionStore : IDisposable
+    {
+        /// <summary>
+        /// Persist the template transaction into the database.
+        /// </summary>
+        /// <param name="templateTransaction">Template transaction to be inserted.</param>
+        Task PutAsync(TemplateTransaction templateTransaction);
+
+        /// <summary>
+        /// Get the template transaction from the database, identified by a hash.
+        /// </summary>
+        /// <param name="hash">A template transaction hash.</param>
+        /// <returns>The template transaction identified by the hash (or null if not found).</returns>
+        Task<TemplateTransaction> GetAsync(uint256 hash);
+
+        /// <summary>
+        /// Delete a template transaction from the database, identified by a hash.
+        /// </summary>
+        /// <param name="hash">Hash of the template transaction to be deleted.</param>
+        Task DeleteAsync(uint256 hash);
+
+        /// <summary>
+        /// Determine if a template transaction already exists in the database.
+        /// </summary>
+        /// <param name="hash">The hash of the template transaction.</param>
+        /// <returns><c>true</c> if the template transaction can be found in the database, otherwise return <c>false</c>.</returns>
+        Task<bool> ExistAsync(uint256 hash);
+    }
+
+    public class TemplateTransactionStore : ITemplateTransactionStore
+    {
+        private const string TemplateTransactionTableName = "TemplateTransaction";
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>Access to DBreeze database.</summary>
+        private readonly DBreezeEngine DBreeze;
+
+        private readonly Network network;
+
+        /// <summary>Provider of time functions.</summary>
+        private readonly IDateTimeProvider dateTimeProvider;
+
+        public TemplateTransactionStore(Network network, DataFolder dataFolder, IDateTimeProvider dateTimeProvider, ILoggerFactory loggerFactory)
+            : this(network, Path.Combine(dataFolder.RootPath, "sidechaindata"), dateTimeProvider, loggerFactory)
+        {
+        }
+
+        public TemplateTransactionStore(Network network, string folder, IDateTimeProvider dateTimeProvider, ILoggerFactory loggerFactory)
+        {
+            Guard.NotNull(network, nameof(network));
+            Guard.NotEmpty(folder, nameof(folder));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            Directory.CreateDirectory(folder);
+            this.DBreeze = new DBreezeEngine(folder);
+            this.network = network;
+            this.dateTimeProvider = dateTimeProvider;
+        }
+
+        /// <summary>Performs any needed initialisation for the database.</summary>
+        public virtual Task InitializeAsync()
```

Comment:
I think they have started using Fody and writing interceptors for this kind of entry and exit logs in the full node, I'll ask for a new package soon and we can do it the same way :)

Reply:
Agreed